src/complex/Parameters/ImportCSVDataParameter.cpp

```diff
@@ -0,0 +1,94 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
+
+#include "ImportCSVDataParameter.hpp"
+
+namespace complex
+{
+// -----------------------------------------------------------------------------
+ImportCSVDataParameter::ImportCSVDataParameter(const std::string& name, const std::string& humanName, const std::string& helpText, const ValueType& defaultValue)
+: ValueParameter(name, humanName, helpText)
+, m_DefaultValue(defaultValue)
+{
+}
+
+// -----------------------------------------------------------------------------
+Uuid ImportCSVDataParameter::uuid() const
+{
+  return ParameterTraits<ImportCSVDataParameter>::uuid;
+}
+
+// -----------------------------------------------------------------------------
+IParameter::AcceptedTypes ImportCSVDataParameter::acceptedTypes() const
+{
+  return {typeid(ValueType)};
+}
+
+// -----------------------------------------------------------------------------
+nlohmann::json ImportCSVDataParameter::toJson(const std::any& value) const
+{
+  const auto& CSVWizardData = GetAnyRef<ValueType>(value);
+  nlohmann::json json = nlohmann::json::object();
+  CSVWizardData.writeJson(json);
+  return json;
+}
+
+// -----------------------------------------------------------------------------
+Result<std::any> ImportCSVDataParameter::fromJson(const nlohmann::json& json) const
+{
+  return {{CSVWizardData::ReadJson(json)}};
+}
+
+// -----------------------------------------------------------------------------
+IParameter::UniquePointer ImportCSVDataParameter::clone() const
+{
+  return std::make_unique<ImportCSVDataParameter>(name(), humanName(), helpText(), m_DefaultValue);
+}
+
+// -----------------------------------------------------------------------------
+std::any ImportCSVDataParameter::defaultValue() const
+{
+  return m_DefaultValue;
+}
+
+// -----------------------------------------------------------------------------
+Result<> ImportCSVDataParameter::validate(const std::any& value) const
+{
+  try
+  {
+    std::any_cast<CSVWizardData>(value);
+  } catch(std::bad_any_cast exception)
+  {
+    return MakeErrorResult(-667, fmt::format("Exception: {}", exception.what()));
+  }
```

Comment:
I have to return a Result<> for this method.

Reply:
The caller of `validate` (which is only `IFilter::preflight`) already checks to make sure that the `std::any` is the right type. So if we get the only type here, we throw an exception which `std::any_cast` will do. Since there is no validation to do on `CSVWizardData` we can just return a default `Result<>`.